src/Microsoft.AspNet.Mvc.Razor/TagHelpers/UrlResolutionTagHelper.cs

```diff
@@ -154,26 +156,45 @@ protected void ProcessUrlAttribute(string attributeName, TagHelperOutput output)
             {
                 foreach (var attribute in attributes)
                 {
-                    string resolvedUrl;
-
                     var stringValue = attribute.Value as string;
                     if (stringValue != null)
                     {
-                        if (TryResolveUrl(stringValue, encodeWebRoot: false, resolvedUrl: out resolvedUrl))
+                        string resolvedUrl;
+                        if (TryResolveUrl(stringValue, resolvedUrl: out resolvedUrl))
                         {
                             attribute.Value = resolvedUrl;
                         }
                     }
                     else
                     {
-                        var htmlStringValue = attribute.Value as HtmlString;
-                        if (htmlStringValue != null &&
-                            TryResolveUrl(
-                                htmlStringValue.ToString(),
-                                encodeWebRoot: true,
-                                resolvedUrl: out resolvedUrl))
+                        var htmlContent = attribute.Value as IHtmlContent;
+                        if (htmlContent != null)
                         {
-                            attribute.Value = new HtmlString(resolvedUrl);
+                            var htmlString = htmlContent as HtmlString;
+                            if (htmlString != null)
+                            {
+                                // No need for a StringWriter in this case.
+                                stringValue = htmlString.ToString();
+                            }
+                            else
+                            {
+                                using (var writer = new StringWriter())
+                                {
+                                    htmlContent.WriteTo(writer, HtmlEncoder);
+                                    stringValue = writer.ToString();
+                                }
+                            }
+
+                            IHtmlContent resolvedUrl;
+                            if (TryResolveUrl(stringValue, resolvedUrl: out resolvedUrl))
+                            {
+                                attribute.Value = resolvedUrl;
+                            }
+                            else if (htmlString == null)
+                            {
+                                // Not a ~/ URL. Just avoid re-encoding the attribute value later.
+                                attribute.Value = new HtmlString(stringValue);
```

Comment:
Since this helper specifically targets known URL-holding attributes, I'm not sure not matching is the "common case". Regardless `attribute.Value` is usually set from a `StringCollectionTextWriter.Content` property and therefore has type `StringCollectionTextWriterContent` (at least today). In that case, saving just a `string` allocation and not re-encoding.

This is more important when the attribute value contains a C# expression of course.